Porting ESRT changes from FreeBSD to DragonflyBSD

kotylamichal opened the pull request.

kotylamichal committed:
Porting ESRT changes from FreeBSD to DragonflyBSD

pcichowski commented:
I reckon that for now, the `get_table_length` and `copy_table` functions aren't needed to print ESRT to dmesg.

This function https://github.com/3mdeb/DragonFlyBSD/blob/e183bc049d7b419f07e9adec0dff67cbfff0106a/sys/platform/pc64/x86_64/efirt.c#L453-L470 is already provided and it allows to get a pointer to a config table with a specified GUID (and the ESRT is just that).

So, I think if you invoked it here, https://github.com/3mdeb/DragonFlyBSD/blob/e183bc049d7b419f07e9adec0dff67cbfff0106a/sys/platform/pc64/x86_64/efirt.c#L435 passing the ESRT guid, you could then cast the void pointer to a ESRT struct pointer and print out its values.

SergiiDmytruk committed:
Copy ESRT in bootloader

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix upstream error in efirt driver

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
Fix bootloader code

Need to find a suitable header for this.  I guess some duplication with
kernel is unavoidable, because including its headers causes compilation
issues.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

kotylamichal committed:
efirt.c: printing ESRT values

SergiiDmytruk commented on stand/boot/efi/loader/arch/x86_64/elf64_freebsd.c:
```diff
@@ -44,6 +44,11 @@
 
+#define	EFI_TABLE_ESRT					\
+	{0xb122a263,0x3661,0x4f68,{0x99,0x29,0x78,0xf8,0xb0,0xd6,0x21,0x80}}
+#define	EFI_PROPERTIES_TABLE			\
+	{0x880aaca3,0x4adc,0x4a04,{0x90,0x79,0xb7,0x47,0x34,0x08,0x25,0xe5}}
```
I know I'm the one who introduced this, but it's not necessary in bootloader here and below unless it doesn't reside in runtime memory. I'm not sure where this table is located and thus whether it needs to be copied. It's 
actually deprecated in UEFI 2.10 and seems to be useless for userspace, so maybe we shouldn't be adding support for it.

And `EFI_TABLE_ESRT` probably needs to be in some header under `stand/`.

kotylamichal replied:
Ok, so I remove all of properties table related code

SergiiDmytruk commented on stand/boot/efi/loader/main.c:
```diff
@@ -51,6 +51,11 @@ extern char bootprog_maker[];
+#define	EFI_TABLE_ESRT					\
+	{0xb122a263,0x3661,0x4f68,{0x99,0x29,0x78,0xf8,0xb0,0xd6,0x21,0x80}}
+#define	EFI_PROPERTIES_TABLE			\
+	{0x880aaca3,0x4adc,0x4a04,{0x90,0x79,0xb7,0x47,0x34,0x08,0x25,0xe5}}
```
Second occurrence is here.

SergiiDmytruk commented on stand/boot/efi/loader/main.c:
```diff
@@ -185,6 +191,38 @@ has_keyboard(void)
 	return retval;
 }
 
+struct efi_esrt_table {
+	uint32_t	fw_resource_count;
+	uint32_t	fw_resource_count_max;
+	uint64_t	fw_resource_version;
+	uint8_t		entries[];
+};
+
+struct efi_esrt_entry_v1 {
+	struct uuid	fw_class;
+	uint32_t 	fw_type;
+	uint32_t	fw_version;
+	uint32_t	lowest_supported_fw_version;
+	uint32_t	capsule_flags;
+	uint32_t	last_attempt_version;
+	uint32_t	last_attempt_status;
+};
```
This should probably be in some header under `stand/`.

kotylamichal replied:
Ok, now it is only in /sys/sys/efi.h

SergiiDmytruk commented on stand/boot/efi/loader/main.c:
```diff
@@ -185,6 +191,38 @@ has_keyboard(void)
+	uint32_t	last_attempt_status;
+};
+
+static void
+efi_set_table(EFI_GUID *tbl, void *value)
+{
+	EFI_GUID *id;
+	unsigned int i;
+
+	for (i = 0; i < ST->NumberOfTableEntries; i++) {
+		id = &ST->ConfigurationTable[i].VendorGuid;
+		if (!memcmp(id, tbl, sizeof(EFI_GUID))) {
+			ST->ConfigurationTable[i].VendorTable = value;
+			return;
+		}
+	}
+}
```
Another of mine test changes, should probably go into `stand/boot/efi/libefi`.

kotylamichal replied:
moved

SergiiDmytruk commented on stand/boot/efi/loader/arch/x86_64/elf64_freebsd.c:
```diff
@@ -96,6 +103,16 @@ elf64_exec(struct preloaded_file *fp)
+	printf("esrt get table done\n");
+	
+	/* end */
+
```
Like on OpenBSD there seems to be multiple copies of bootloader and need to figure out with upstream whether all of them need to be updated or only one is actually used.

kotylamichal replied:
@SergiiDmytruk That was redundant changes - now `elf64_freebsd.c` has no differences. Of course in the upstream phase I will ask community about correct copy of bootloader

SergiiDmytruk commented on stand/boot/efi/loader/main.c:
```diff
@@ -339,6 +377,31 @@ main(int argc, CHAR16 *argv[])
+		EFI_STATUS status = BS->AllocatePool(EfiRuntimeServicesData,
+								  esrt_size, &esrt_copy);
+		if (status != EFI_SUCCESS)
+			printf("BS->AllocatePool() has failed");
```
This needs to be handled better, with abort or `else` branch, because otherwise `NULL` or random pointer will be used for `memcpy()` below.

kotylamichal replied:
got it

SergiiDmytruk commented on sys/platform/pc64/x86_64/efirt.c:
```diff
@@ -464,6 +501,130 @@ efi_get_table(struct uuid *uuid, void **ptr)
+
+		buf = kmalloc(len, M_TEMP, M_WAITOK);
+		//Temporary solution - maybe copyout instead of physcopyout
+		//error = physcopyout((vm_paddr_t)esrt, buf, len);
```
You might be able to copy `physcopyout` from https://github.com/freebsd/freebsd-src/blob/373ffc62c158e52cde86a5b934ab4a51307f9f2e/sys/kern/subr_uio.c#L119-L138

Otherwise need to temporary map the table into virtual address space: `bus_space_map` + `bus_space_vaddr` + `copyout` + `bus_space_unmap`. This can serve as an implementation for `physcopyout`.

SergiiDmytruk reviewed (commented):
This targets release from January 2022, you might want to try rebasing onto current master branch. There might be no conflicts, but it's better to test that it still works there and also necessary for sending patches upstream as they might not apply otherwise.

kotylamichal committed:
ESRT tables: fixes after review

kotylamichal commented:
@SergiiDmytruk I make some fixes after your review

SergiiDmytruk commented on stand/boot/efi/loader/main.c:
```diff
@@ -395,11 +358,12 @@ main(int argc, CHAR16 *argv[])
 		EFI_STATUS status = BS->AllocatePool(EfiRuntimeServicesData,
 								  esrt_size, &esrt_copy);
-		if (status != EFI_SUCCESS)
+		if (status != EFI_SUCCESS){
```
Inverted condition.

kotylamichal replied:
fixed

SergiiDmytruk commented on stand/boot/efi/libefi/libefi.c:
```diff
@@ -48,6 +48,21 @@ efi_get_table(EFI_GUID *tbl)
 	return (NULL);
 }
 
+static void
```
`static`? You won't be able to use it then.

kotylamichal replied:
fixed

SergiiDmytruk reviewed (changes requested):
EFI bootloader doesn't seem to be in a working state.

kotylamichal committed:
efi: fixes after review

kotylamichal committed:
fix

kotylamichal committed:
fixes

kotylamichal committed:
cleanup

kotylamichal committed:
cleanup

kotylamichal committed:
fixes

kotylamichal commented:
@SergiiDmytruk Now should be better, I make some cleanups and rebuild bootloader and kernel - everything works properly

SergiiDmytruk commented on sys/sys/efi.h:
```diff
@@ -40,7 +40,7 @@
 	{0xeb9d2d32,0x2d88,0x11d3,0x9a,0x16,{0x00,0x90,0x27,0x3f,0xc1,0x4d}}
 #define	EFI_TABLE_ESRT					\
-	{0xb122a263,0x3661,0x4f68,0x99,0x29,{0x78,0xf8,0xb0,0xd6,0x21,0x80}}
+	{0xb122a263,0x3661,0x4f68,{0x99,0x29,0x78,0xf8,0xb0,0xd6,0x21,0x80}}
```
This change and `#undef`s above is why I was talking about header under `stand/`. They probably don't include `sys/` headers on purpose here. ioctl for ESRT likely won't work with this initializer because of `uuid_t`.

Maybe we should just add https://github.com/tianocore/edk2/blob/master/MdePkg/Include/Guid/SystemResourceTable.h to `sys/contrib/edk2` and include it? I suspect that might be the solution upstream would go for.

kotylamichal replied:
@SergiiDmytruk I tried with that, but there are a lot of errors with unknown types like `UINT32` and `UINT64`. We should define these macros or include more from edk2, but I'm not sure that really is a more "clean" solution than that undef. Maybe it is enough for this time, and I should ask the community on the ML (at the upstream phase) which solution will be acceptable?

> ioctl for ESRT likely won't work with this initializer because of uuid_t

Ok, but ioctl is in the scope of the next task. It will be in different PR and I will still work on that

SergiiDmytruk replied:
> Maybe it is enough for this time, and I should ask the community on the ML (at the upstream phase) which solution will be acceptable?

OK, I'm fine with that.

kotylamichal replied:
@SergiiDmytruk We have one discission left here - can you resolve that?

SergiiDmytruk replied:
Resolved.
